Adds documentation about how to configure CMO (#1696)
openshift/cluster-monitoring-operator: Documentation/user-guides/configuring-cluster-monitoring.md

# Boolean flag to enable monitoring for user-defined projects.
enableUserWorkload: <bool>
# Holds configuration related with prometheus-adapter, nodeSelector and tolerations
# and Audit level
k8sPrometheusAdapter: <struct>
# Holds configuration related with kube-state-metrics agent, nodeSelector and tolerations
kubeStateMetrics: <struct>
# Holds configuration related with the prometheus component that will be managed by
# the prometheus-operator from configuring nodeSelector to RemoteWrite
prometheusK8s: <struct>
# Holds configuration related with prometheus-operator, nodeSelector, tolerations
# and logLevel
prometheusOperator: <struct>
# Holds configuration related with openshift-state-metrics agent, nodeSelector and tolerations
openshiftStateMetrics: <struct>
# Holds configuration related with the Thanos Querier component, nodeSelector, tolerations
thanosQuerier: <struct>
```

### alertmanagerMain

Holds configuration for all things Alertmanager

```yaml
alertmanagerMain:
  # Boolean flag to enable or disable the alertmanager instance that
  # will live under openshift-monitoring
  enabled: <bool>
  # Boolean flag to enable or disable the alertmanager instance for

>   # user workload monitoring  under openshift-user-workload-monitoring

Reviewer: this enables the reconciliation of `AlertmanagerConfig` resources but not the dedicated UWM alertmanager.

  enableUserAlertmanagerConfig: <bool>
  # Log level for Alertmanager to be configured with.
  # Possible values are: error, warn, info, debug.
  logLevel: <string>
  # Define which Nodes the Pods are scheduled on.
  nodeSelector: v1.NodeSelector
  # Define resources requests and limits for single Pods.
  resources: v1.ResourceRequirements
  # Defines the Pods tolerations.
  tolerations: v1.Toleration
  # Allows configuring persistent storage for the Alertmanager Pods
  # it's possible to configure storageClass and size of volume.
  volumeClaimTemplate: v1.PersistentVolumeClaim
```

Example

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    alertmanagerMain:
      enabled: true
      enableUserAlertmanagerConfig: true
      logLevel: "debug"
      nodeSelector:
        kubernetes.io/os: linux
      tolerations:
        - operator: Exists
      resources:
        requests:
          cpu: 5m

>           memory: 30Mi

Reviewer: can you also include `volumeClaimTemplate` in the example?

Example

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    k8sPrometheusAdapter:
      audit:
        profile: request
      nodeSelector:
        kubernetes.io/os: linux
      tolerations:
        - operator: Exists
```

### kubeStateMetrics

Holds configuration for all things kube-state-metrics agent.

```yaml
kubeStateMetrics:
  # Define which Nodes the Pods are scheduled on.
  nodeSelector: v1.NodeSelector
  # Defines the Pods tolerations.
  tolerations: v1.Toleration
```

Example

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    kubeStateMetrics:
      nodeSelector:
        kubernetes.io/os: linux
      tolerations:
        - operator: Exists
```

### prometheusK8s

Holds configuration related to the Prometheus component that will be managed by prometheus-operator.

```yaml
prometheusK8s:

>   # Holds configuration about how the Prometheus component should communicate with Alertmanager,

Reviewer: ```suggestion
  # Holds configuration about how the Prometheus component should communicate with additional Alertmanager instances,
```

  # from configuring the URL scheme, timeout to authentication method and others.
  additionalAlertmanagerConfigs: <struct>
  # Enforces body size limit of Prometheus scrapes, if a scrape is bigger than
  # the limit it will fail.
  # 3 kinds of values are accepted:
  #  1. empty value: no limit
  #  2. a value in Prometheus size format, e.g. "64MB"
  #  3. string "automatic", which means the limit will be automatically calculated based on
  #     cluster capacity.
  enforcedBodySizeLimit: <string>
  # The labels to add to any time series or alerts when communicating with external systems
  # (federation, remote storage, Alertmanager).
  externalLabels: <struct>
  # Log level for Prometheus to be configured with.
  # Possible values are: error, warn, info, debug.
  logLevel: <string>
  # Define which Nodes the Pods are scheduled on.
  nodeSelector: v1.NodeSelector
  # QueryLogFile specifies the file to which PromQL queries are logged. Suports both just a filename
  # in which case they will be saved to an emptyDir volume at /var/log/prometheus, if a full path is

>   # given an emptyDir volume will be mounted at that location.

Reviewer: I think that the rules are more complicated than that so we need to be more exhaustive.

Author: Humm I'm not quite sure what I might be missing. I've added regarding relative paths and linux std streams. Let me know if I'm missing something

  queryLogFile: <string>
  # Holds the remote write configuration, everything from url, authorization to relabeling
  remoteWrite: <struct>
  # Define resources requests and limits for single Pods.
  resources: v1.ResourceRequirements
  # Time duration Prometheus shall retain data for. Default is '24h' if retentionSize is not set,
  # and must match the regular expression [0-9]+(ms|s|m|h|d|w|y) (milliseconds seconds minutes
  # hours days weeks years).
  retention: <string>

>   # Maximum amount of disk space used by blocks.

Reviewer: IIRC it's blocks + WAL in fact which is to say all the data.

Author: I took this from p-o docs, but I double checked and you are correct 👌

  # in the namespace where it was deployed to
  # Default: false
  enableAlertmanagerConfig: <bool>
  # Log level for Alertmanager to be configured with.
  # Possible values are: error, warn, info, debug.
  logLevel: <string>
  # Define which Nodes the Pods are scheduled on.
  nodeSelector: v1.NodeSelector
  # Define resources requests and limits for single Pods.
  resources: v1.ResourceRequirements
  # Defines the Pods tolerations.
  tolerations: v1.Toleration
  # Allows configuring persistent storage for the Alertmanager Pods
  # it's possible to configure storageClass and size of volume.
  volumeClaimTemplate: v1.PersistentVolumeClaim
```

Example:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    alertmanager:
      enabled: true
      enableAlertmanagerConfig: true
      logLevel: "debug"
      nodeSelector:
        kubernetes.io/os: linux
      tolerations:
        - operator: Exists
      resources:
        requests:
          cpu: 5m
          memory: 30Mi
```

### prometheus

> Holds configuration related to the Prometheus component that will be managed by prometheus-operator and monitor user-defined projects.

Reviewer: ```suggestion
Holds configuration related to the Prometheus component that will monitor user-defined projects.
```

```yaml
prometheus:

>   # Holds configuration about how the Prometheus component should communicate with Alertmanager,

Reviewer: ```suggestion
  # Holds configuration about how the Prometheus component should communicate with additional Alertmanager instances,
```

  # from configuring the URL scheme, timeout to authentication method and others.
  additionalAlertmanagerConfigs: <struct>
  # Per-scrape limit on the number of labels accepted for a sample. If more than this
  # number of labels are present post metric-relabeling, the entire scrape will be treated as

>   # failed. 0 means no limit.

Reviewer: ```suggestion
  # failed. 0 means no limit.
  # default: 0
```

  enforcedLabelLimit: <integer>
  # Per-scrape limit on the length of labels name that will be accepted for a sample. If a label name
  # is longer than this number post metric-relabeling, the entire scrape will be treated as

>   # failed. 0 means no limit.

Reviewer: ditto

  enforcedLabelNameLengthLimit: <integer>
  # Per-scrape limit on the length of labels value that will be accepted for a sample. If a label
  # value is longer than this number post metric-relabeling, the entire scrape will be treated

>   # as failed. 0 means no limit.

Reviewer: ditto

  enforcedLabelValueLengthLimit: <integer>
  # EnforcedSampleLimit defines a global limit on the number of scraped samples that will be accepted.
  # This overrides any SampleLimit set per ServiceMonitor or/and PodMonitor. It is meant to be
  # used by admins to enforce the SampleLimit to keep the overall number of samples/series under the

>   # desired limit. Note that if SampleLimit is lower that value will be taken instead.

Reviewer: ditto

The configuration file itself is always defined under the `config.yaml` key within the ConfigMap's data.

Monitoring a platform such as OpenShift requires a coordination of multiple components that must work well between themselves.
However, users should be able to customize the monitoring stack in such a way that they end up with a resilient and highly available monitoring solution.
Despite this, to avoid users from misconfiguring the monitoring stack of their clusters not all configuration parameters are exposed.

Configuring Cluster Monitoring is optional. If the config does not exist or is empty or malformed, then defaults will be used.

## Index

- [Configuration of OpenShift Container Platform](#configuration-of-openShif-container-platform)
  - [alertmanagerMain](#alertmanagerMain)
  - [enableUserWorkload](#enableuserworkloadmonitoring)
  - [k8sPrometheusAdapter](#k8sprometheusadapter)
  - [kubeStateMetrics](#kubeStateMetrics)
  - [prometheusK8s](#prometheusK8s)
  - [prometheusOperator](#prometheusOperator)
  - [openshiftStateMetrics](#openshiftStateMetrics)

>   - [thanosQuerier](#thanosQuerier)

Reviewer 2: missed `telemeterClient`

Author: Yes, I didn't include it as it doesn't concern user configuration

  # instance is not enabled
  # default: false
  enableUserAlertmanagerConfig: <bool>
  # Log level for Alertmanager to be configured with.
  # Possible values are: error, warn, info, debug.
  # default: info
  logLevel: <string>
  # Define which Nodes the Pods are scheduled on.
  nodeSelector: v1.NodeSelector
  # Define resources requests and limits for single Pods.
  resources: v1.ResourceRequirements
  # Defines the Pods tolerations.
  tolerations: v1.Toleration
  # Allows configuring persistent storage for the Alertmanager Pods
  # it's possible to configure storageClass and size of volume.
  volumeClaimTemplate: v1.PersistentVolumeClaim
```

Example

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    alertmanagerMain:
      enabled: true
      enableUserAlertmanagerConfig: true
      logLevel: "debug"
      nodeSelector:
        kubernetes.io/os: linux
      tolerations:
        - operator: Exists
      resources:
        requests:
          cpu: 5m
          memory: 30Mi
      volumeClaimTemplate:
        spec:
          resources:
            requests:

>               storage: 15Gb

Reviewer 2: set `storage: 15Gb` will see error in CMO
```
E0701 12:22:49.300250       1 operator.go:544] sync "openshift-monitoring/cluster-monitoring-config" failed: the Cluster Monitoring ConfigMap could not be parsed: error unmarshaling JSON: while decoding JSON: quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'
```
set to 15Gi or 15G

metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    enableUserWorkload: true
```

### k8sPrometheusAdapter

Holds configuration for all things Prometheus adapter.

```yaml
k8sPrometheusAdapter:
  # Audit configuration to be used by the prometheus adapter instance,
  # possible profile values are: "metadata, request, requestresponse, none".
  # default: metadata
  audit: <struct>
  # Define which Nodes the Pods are scheduled on.
  nodeSelector: v1.NodeSelector
  # Defines the Pods tolerations.
  tolerations: v1.Toleration
```

For more in-depth documentation on how to configure audit please refer to [OpenShift Docs setting audit log levels for prometheus adapter](https://docs.openshift.com/container-platform/latest/monitoring/configuring-the-monitoring-stack.html#setting-audit-log-levels-for-the-prometheus-adapter_configuring-the-monitoring-stack).

Example

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    k8sPrometheusAdapter:
      audit:
        profile: request
      nodeSelector:
        kubernetes.io/os: linux
      tolerations:

>         - operator: Exists

Reviewer 2: ```
      tolerations:
      - operator: Exists
```

Reviewer 2: same for other similar tolerations setting

  volumeClaimTemplate: v1.PersistentVolumeClaim
```

Example

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    prometheus:
      additionalAlertmanagerConfigs:
      - apiVersion: v2
        scheme: https
        bearerToken:
          name: alertmanager1-bearer-token
          key: token
        staticConfigs:
        - alertmanager1-remote.com
      enforcedLabelLimit: 500
      enforcedLabelNameLengthLimit: 50
      enforcedLabelValueLengthLimit: 600
      enforcedSampleLimit: 100
      enforcedTargetLimit: 10
      externalLabels:
        datacenter: eu-west
      logLevel: debug
      nodeSelector:
        kubernetes.io/os: linux
      queryLogFile: /tmp/test.log
      retention: 10h
      resources:
        requests:
          cpu: 100m
          memory: 100Mi
      remoteWrite:
      - url: "https://test.remotewrite.com/api/write"
      retention: 24h
      retentionSize: 15GB
      tolerations:
        - operator: "Exists"
      volumeClaimTemplate:
        spec:
          resources:
            requests:

>               storage: 15Gb

Reviewer 2: 15Gi or 15G

  # and must match the regular expression [0-9]+(ms|s|m|h|d|w|y) (milliseconds seconds minutes
  # hours days weeks years).
  # default: 15d
  retention: <string>
  # Defines the Pods tolerations.
  tolerations: v1.Toleration
  # Allows configuring persistent storage for the Alertmanager Pods
  # it's possible to configure storageClass and size of volume.
  volumeClaimTemplate: v1.PersistentVolumeClaim
```

Example:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    thanosRuler:
      additionalAlertmanagerConfigs:
      - apiVersion: v2
        scheme: https
        bearerToken:
          name: alertmanager1-bearer-token
          key: token
        staticConfigs:
        - alertmanager1-remote.com
      logLevel: debug
      nodeSelector:
        kubernetes.io/os: linux
      resources:
        requests:
          cpu: 100m
          memory: 100Mi
      retention: 24h
      tolerations:
        - operator: "Exists"
      volumeClaimTemplate:
        spec:
          resources:
            requests:

>               storage: 15Gb

Reviewer 2: 15Gi or 15G

```

### RemoteWriteSpec

Holds configuration for remote write

```yaml
remoteWrite:
  # The URL of the endpoint to send samples to.
  url: <string>
  # The name of the remote write queue, must be unique if specified. The
  # name is used in metrics and logging in order to differentiate queues.
  # Only valid in Prometheus versions 2.15.0 and newer.
  name: <string>
  # Timeout for requests to the remote write endpoint.
  remoteTimeout: <string>
  # Custom HTTP headers to be sent along with each remote write request.
  # Be aware that headers that are set by Prometheus itself can't be overwritten.
  # Only valid in Prometheus versions 2.25.0 and newer.
  headers: <map[string]string>
  # The list of remote write relabel configurations.
  writeRelabelConfigs: RelabelConfig
  # BasicAuth for the URL.
  basicAuth: BasicAuth
  # Bearer token for remote write.
  bearerTokenFile: <string>
  # TLS Config to use for remote write.
  tlsConfig: SafeTLSConfig
  # Optional ProxyURL
  proxyUrl: <string>
  # QueueConfig allows tuning of the remote write queue parameters.
  queueConfig: QueueConfig
  # MetadataConfig configures the sending of series metadata to remote storage.

>   metadataConfig: MetadataConfig

Reviewer 2: from https://github.com/openshift/cluster-monitoring-operator/blob/release-4.11/pkg/manifests/config.go#L155-L188
there are also `authorization`, `sigv4`, `oauth2`

Author: Yes, I didn't include them since this was mostly to fulfil user documentation for 4.10. Once @bburt-rh is finished I'll add those missing fields

```

#### BasicAuth

```yaml
basicAuth:
  # The secret in the service monitor namespace that contains the username
  # for authentication.
  username: v1.SecretKeySelector
  # The secret in the service monitor namespace that contains the password
  # for authentication.
  password: v1.SecretKeySelector
```

#### MetadataConfig

```yaml
metadataConfig:
  # Whether metric metadata is sent to remote storage or not.
  send: <bool>
  # How frequently metric metadata is sent to remote storage.
  sendInterval: <string>
```

#### QueueConfig

```yaml
queueConfig:
  # BatchSendDeadline is the maximum time a sample will wait in buffer.
  batchSendDeadline: <string>
  # Capacity is the number of samples to buffer per shard before we start dropping them.
  capacity: <integer>
  # MaxBackoff is the maximum retry delay.
  maxBackoff: <string>
  # MaxRetries is the maximum number of times to retry a batch on recoverable errors.

>   maxRetries: <integer>

Reviewer 2: from https://github.com/openshift/cluster-monitoring-operator/blob/release-4.11/vendor/github.com/prometheus/prometheus/config/config.go#L834-L855
there is not `maxRetries`

# Configuring Cluster Monitoring

> Parts of Cluster Monitoring are configurable. Depending which part of the stack the user wants to configure, he should edit the following: 

Reviewer: ```suggestion
Parts of Cluster Monitoring are configurable. Depending which part of the stack users want to configure, they should edit the following: 
```

> - core infrastruture monitoring configuration lies in a ConfigMap called `cluster-monitoring-config` in the `openshift-monitoring` namespace.

Reviewer: ```suggestion
- Configuration of OpenShift Container Platform monitoring components lies in a ConfigMap called `cluster-monitoring-config` in the `openshift-monitoring` namespace.
```

> - user workload monitoring configuration lies in a ConfigMap called `user-workload-monitoring-config` in the `openshift-user-workload-monitoring` namespace.

Reviewer: ```suggestion
- Configuration of components that monitor user-defined projects lies in a ConfigMap called `user-workload-monitoring-config` in the `openshift-user-workload-monitoring` namespace.
```

The configuration file itself is always defined under the `config.yaml` key within the ConfigMap's data.

Configuring Cluster Monitoring is optional. If the config does not exist, or is empty or malformed, then defaults will be used.

#### Note

> For simplicity in the following sections we will ommit the yaml necessary for the configuration to be persisted so for the example:

Reviewer: ```suggestion
For simplicity in the following sections we will omit the yaml necessary for the configuration to be persisted so for the example:
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    <FIELD_TO_BE_CONFIGURABLE>: <VALUE>
```
We will only show:
```yaml
<FIELD_TO_BE_CONFIGURABLE>: <VALUE>
```

> ## Core infrastruture monitoring

Reviewer 3: nit: Core Infrastructure Monitoring

Reviewer: the OCP docs say " Core OpenShift Container Platform monitoring components"

Monitoring a platform such as OpenShift requires a coordination of multiple components that must work well between themselves.
However, users should be able to customize the monitoring stack in such a way that they end up with a resilient and highly available monitoring solution.

> Despite this, to avoid users from missconfiguring the monitoring stack of their clusters not all configuration parameters are exposed.

Reviewer 3: nit: missconfiguring -> misconfiguring

The ConfigMap `cluster-monitoring-config` in the `openshift-monitoring` namespace is the root for all configuration of the core infrastruture monitoring. Here we can configure the following:

```yaml

> # Holds configuration for everything Alertmanager related in the cluster

Reviewer: ```suggestion
# Holds configuration for the Alertmanager component in the cluster
```

# from configuring nodeSelectors to enabeling UserAlertManagerConfig
alertmanagerMain: <struct>

> # Boolean flag to enable UserWorkloadMonitoring

Reviewer: ```suggestion
# Boolean flag to enable monitoring for user-defined projects.
```

enableUserWorkload: <bool>
# Holds configuration related with prometheus-adapter, nodeSelector and tolerations
# and Audit level
k8sPrometheusAdapter: <struct>

> # Holds configuration related with kubeStateMetrics, nodeSelector and tolerations

Reviewer: ```suggestion
# Holds configuration related with kube-state-metrics agent, nodeSelector and tolerations
```

kubeStateMetrics: <struct>

> # Holds configuration related with the prometheus pods that will be managed by

Reviewer: ```suggestion
# Holds configuration related with the prometheus component that will be managed by
```

# the prometheus-operator from configuring nodeSelector to RemoteWrite
prometheusK8s: <struct>
# Holds configuration related with prometheus-operator, nodeSelector, tolerations 
# and logLevel 
prometheusOperator: <struct>

> # Holds configuration related with openshift-state-metrics, nodeSelector and tolerations

Reviewer: ```suggestion
# Holds configuration related with openshift-state-metrics agent, nodeSelector and tolerations
```

openshiftStateMetrics: <struct>

> # Holds configuration related with thanos-querier, nodeSelector, tolerations 

Reviewer: ```suggestion
# Holds configuration related with the Thanos Querier component, nodeSelector, tolerations 
```

thanosQuerier: <struct>
```

### alertmanagerMain

> Holds configuration for all things alermanager

Reviewer 3: nit: alertmanager

```yaml
alertmanagerMain:

>   # Boolean flag to enable or disable the alertmanager instance that 

Reviewer: ```suggestion
  # Boolean flag to enable or disable the Alertmanager instance that 
```

  # will live under openshift-monitoring 
  enabled: <bool>
  # Boolean flag to enable or disable the alertmanager instance for 
  # user workload monitoring  under openshift-user-workload-monitoring
  enableUserAlertmanagerConfig: <bool>
  # Log level to be used by the main alertmanager instance, possible values are

>   # info, warn, debug. #TODO validate this

Reviewer: ```suggestion
  # error, warn, info, debug.
  # default: info
```

  logLevel: <string>
  # Define which Nodes the Pods are scheduled on.
  nodeSelector: <struct>
  # Defines the alertmanager Pods tolerations.
  tolerations: <struct>
  # Define resources requests and limits for alertmanager Pods.
  resources: <struct>
  # Allows configuring persistent storage for the alertmanager Pods 
  # users can configure storageClass and size of volume.
  volumeClaimTemplate: <struct>
```

Example

```yaml
alertmanagerMain:
  enabled: true
  enableUserAlertmanagerConfig: true
  logLevel: "debug"
  nodeSelector:
    kubernetes.io/os: linux
  tolerations:
    - operator: Exists
  resources:
    requests:
      cpu: 5m
      memory: 30Mi
```

### enableUserWorkloadMonitoring

> Enabels or disabels user workload monitoring, by default is `false`. When enabeled, CMO will create a new namespace `openshift-user-workload-monitoring` to where it will deploy the following components:

Reviewer 3: nit: Enables or disables by default disabled

When enabled CMO will create a new namespace  called `openshift-user-workload-monitoring` to which it will deploy the following components

> Enabels or disabels user workload monitoring, by default is `false`. When enabeled, CMO will create a new namespace `openshift-user-workload-monitoring` to where it will deploy the following components:

Reviewer: ```suggestion
Enables (when true) or disables user workload monitoring, by default is `false`. When enabled, CMO will create a new namespace `openshift-user-workload-monitoring` to where it will deploy the following components:
```

- prometheus-operator
- prometheus
- thanos-ruler 

> These components will be responsible for monitoring all namespaces which do not have the label `openshift.io/cluster-monitoring: true` or `openshift.io/user-monitoring: false`.

Reviewer: The OCP docs don't explain which projects are considered to be user-defined [1]. The reference documentation should probably be consistent (or link to [1]?).

[1] https://docs.openshift.com/container-platform/4.10/monitoring/enabling-monitoring-for-user-defined-projects.html

```yaml

> enableUserWorkload: <bool>

Reviewer 3: I think its better to add the entire config map than just key?

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    enableUserWorkload: <bool>
```

```

Example

```yaml

> enableUserWorkload: true

Reviewer 3: ```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    enableUserWorkload: true
```

```

### k8sPrometheusAdapter

Holds configuration for all things prometheus adapter.

```yaml
k8sPrometheusAdapter:
  # Audit configuration to be used by the prometheus adapter instance, 
  # possible profile values are: "metadata, request, requestresponse, none".

>   # The default audit log level is "metadata"

Reviewer: For the record, we already have https://docs.openshift.com/container-platform/4.10/monitoring/configuring-the-monitoring-stack.html#setting-audit-log-levels-for-the-prometheus-adapter_configuring-the-monitoring-stack

  audit: <struct>
  # Define which Nodes the Pods are scheduled on.
  nodeSelector: <struct>

>   # Defines the alertmanager Pods tolerations.

Reviewer: ```suggestion
  # Defines the Pods tolerations.
```

  tolerations: <struct>
```

Example

```yaml
k8sPrometheusAdapter:
  audit:
    profile: request
  nodeSelector:
    kubernetes.io/os: linux
  tolerations:
    - operator: Exists
```

## User workload monitoring

> 

Reviewer 3: do we need blank lines here?
